### ISSUE TITLE
Focused terminal uses CPU-heavy DOM renderer instead of WebGL

### ISSUE THREAD

**Issue by gregpriday** [canopyide/canopy#3156]
## Summary

All terminal panels render exclusively through the xterm.js DOM renderer, which creates thousands of `<span>` and `<div>` elements per frame and is CPU-bound. Since xterm.js 6.0 removed the canvas renderer, WebGL is the only hardware-accelerated rendering path available. For the active/focused terminal — the one the user is watching during a build, test run, or agent task — switching to WebGL would substantially reduce the CPU cost of high-throughput output rendering.

## Problem Statement

The DOM renderer works reliably for any number of terminals and avoids the Chromium-enforced limit of 16 active WebGL contexts per renderer process. This was a reasonable default choice. However, the trade-off is significant for the terminal the user is actually watching:

- The DOM renderer allocates one DOM node per terminal cell for rendered rows. For a 220-column × 50-row terminal, that's 11,000+ nodes that must be diffed and repainted on each output frame.
- During high-throughput output (build logs, test runners, `cat` of a large file), the focused terminal drives repeated full-viewport reflows that push CPU usage noticeably higher than necessary.
- The WebGL renderer (`@xterm/addon-webgl`) renders the entire terminal viewport as a single `<canvas>` element, offloading character rasterisation and compositing to the GPU. Benchmarks show WebGL is up to 900% faster than the canvas renderer (now removed in 6.0), which itself was 5–45× faster than the DOM renderer.
- With only one WebGL context needed (the focused terminal), the 16-context per-process ceiling is not approached, and the existing DOM renderer continues serving all other panels.

## Rendering Options in xterm.js 6.0

xterm.js 6.0 removed `@xterm/addon-canvas` (the 2D Canvas renderer). The only two rendering paths available are:

| Renderer | How it works | Pros | Cons |
|----------|-------------|------|------|
| **DOM** (built-in) | One DOM node per cell, diffed on each frame | Always available, no GPU required, unlimited instances | CPU-bound, layout thrashing on high throughput |
| **WebGL** (`@xterm/addon-webgl`) | Single `<canvas>`, GPU-composited draw calls | Orders of magnitude faster for CPU path | Limited to 16 contexts per Chromium renderer process, requires GPU |

There is no middle-ground canvas fallback. The architecture must use DOM as the universal fallback and WebGL as the performance tier.

## Desired Behavior

The terminal with active keyboard focus and/or the terminal currently in the foreground renderer tier should use the WebGL addon for hardware-accelerated rendering. All non-focused panels continue using the DOM renderer.

When the focused terminal changes, the WebGL context is detached from the old terminal and attached to the new one. If WebGL context creation fails for any reason (GPU process crash, context loss event from `onContextLoss`, or the `CANOPY_DISABLE_WEBGL` flag), the terminal transparently falls back to the DOM renderer without any user-visible disruption.

### WebGL Lifecycle

1. When a terminal gains the `FOCUSED` tier (via `TerminalRendererPolicy`), attempt to attach `@xterm/addon-webgl`.
2. Before attaching to the new terminal, fully `.dispose()` the WebGL addon on the previously focused terminal. The old terminal reverts to the DOM renderer automatically.
3. Wrap `terminal.loadAddon(webglAddon)` in a `try/catch`. If it throws (no GPU, software rendering, etc.), silently remain on the DOM renderer.
4. Subscribe to the addon's `onContextLoss` event. On context loss, `.dispose()` the addon immediately — the terminal falls back to DOM. WebGL can be re-attempted on the next focus change.
5. Debounce focus changes to prevent rapid attach/dispose cycles (the existing `TIER_DOWNGRADE_HYSTERESIS_MS` in `TerminalRendererPolicy` already provides this for downgrades; upgrades to `FOCUSED` are immediate but should be debounced for WebGL attachment specifically).

### `CANOPY_DISABLE_WEBGL` Environment Flag

Add a `CANOPY_DISABLE_WEBGL` environment variable that completely disables WebGL addon loading when set to `"1"`. This serves three purposes:

1. **E2E testing** — CI runners pass `--disable-gpu` and have no real GPU. While the `try/catch` fallback would handle this, explicitly disabling WebGL avoids unnecessary addon loading overhead and makes test behavior deterministic.
2. **Debugging** — Developers can isolate whether a rendering issue is WebGL-related by running with this flag.
3. **User escape hatch** — Users on problematic GPU drivers can force DOM-only mode.

The flag should be checked in `TerminalAddonManager` before any WebGL attachment attempt.

### E2E Test Integration

The e2e test suite reads terminal content via `.xterm-rows` `innerText()` (see `e2e/helpers/terminal.ts`). While xterm.js keeps `.xterm-rows` populated for accessibility even with WebGL active, explicitly disabling WebGL in tests ensures deterministic behavior and avoids any risk of test flakiness.

In `e2e/helpers/launch.ts`, add to the CI launch environment:

```typescript
CANOPY_DISABLE_WEBGL: "1",
```

This should be set for **all** e2e test projects (core and online), not just online tests, since any test that reads terminal content or interacts with terminals benefits from deterministic DOM rendering.

## Context

The rendering tier system and focus tracking already exist in the codebase and provide the signals needed to know which terminal is "active":

- [`src/services/terminal/TerminalRendererPolicy.ts`](https://github.com/canopyide/canopy/blob/develop/src/services/terminal/TerminalRendererPolicy.ts) — tier management (`FOCUSED` → `VISIBLE` → `BACKGROUND`) with hysteresis
- [`src/services/terminal/TerminalAddonManager.ts`](https://github.com/canopyide/canopy/blob/develop/src/services/terminal/TerminalAddonManager.ts) — where addon loading happens; line 23 explains the explicit decision to avoid WebGL context limits

The `@xterm/addon-webgl` package is already a transitive dependency of `@xterm/xterm` 6.0 (confirmed in `package.json`). The WebGL addon exposes an `onContextLoss` event specifically designed for graceful degradation.

## Acceptance Criteria

- [ ] The focused/active terminal uses `@xterm/addon-webgl` for rendering; all other terminals continue using the DOM renderer
- [ ] When focus shifts to a different terminal, the WebGL context transfers to that terminal (or a new one is allocated) within the existing tier-change hysteresis window
- [ ] A `webglcontextlost` event on the focused terminal's canvas causes immediate, silent fallback to the DOM renderer for that terminal; focus can be regained and WebGL re-attempted on the next focus change
- [ ] No visual artefacts (flash, blank frame, cursor jump) occur during WebGL attach or detach transitions
- [ ] The total active WebGL context count remains 1 at steady state
- [ ] Setting `CANOPY_DISABLE_WEBGL=1` completely prevents WebGL addon loading and all terminals use the DOM renderer
- [ ] All e2e tests (core and online) run with `CANOPY_DISABLE_WEBGL=1` set in the launch environment
- [ ] The existing test suite and E2E coverage pass without new failures

## Edge Cases & Risks

- **Rapid focus changes** (tab switching, keyboard shortcut cycling) must not create multiple leaked WebGL contexts. The old addon must be fully disposed before attaching a new one. Focus-change debouncing (~50-100ms) should gate WebGL attachment.
- **GPU failures** — Some GPU configurations (software-only rendering in VMs, certain Linux setups) will fail WebGL context creation silently. The `try/catch` on `loadAddon` and the `onContextLoss` handler are both required for correct fallback.
- **Disposal cleanup** — Calling `.dispose()` on the WebGL addon is sufficient to remove the injected canvas and free GPU resources (confirmed for xterm.js 6.0). No manual DOM manipulation is needed.
- **Known xterm.js bugs** — A GPU memory leak on disposal was fixed in v5.0 (Issue #3889), and a race condition during terminal disposal involving the WebGL addon was reported in Oct 2024 (Issue #5181). Ensure disposal is sequential, not concurrent.
- **Shader compilation latency** — The first WebGL attachment after app launch compiles shaders, which may cause a brief delay. Subsequent attach/dispose cycles reuse compiled shaders if the GPU process hasn't been recycled. This is acceptable.

**Comment by gregpriday** [comment]
Dug through the git history on this. We've actually had three rounds of GPU-accelerated rendering, all on xterm.js 5.x, and the full arc is worth understanding before we take another shot.

## What we tried before

Back in December 2025, there was a full `WebGLContextManager` (227 lines) with dynamic context budgets (4-16 based on CPU cores), visibility-aware LRU eviction, exponential backoff recovery, grace periods for hidden terminals, and IntersectionObserver integration. The key commits tell the story:

- `8d1bc9b8` — dynamic WebGL budget based on hardware detection
- `023c9f1e` — visibility-aware context management with IntersectionObserver
- `7e05c769` — TUI flashing fix with atomic frame updates and backoff recovery
- `d825b264` — off-by-one fix in eviction boundary (the kind of bug that tells you the system is too complex)

The problems kept stacking up: rendering artifacts after drag operations, TUI flashing, context loss during rapid focus changes, and rendering glitches we couldn't fully pin down. WebGL was disabled on Dec 14 (`ead7c529`), replaced briefly with Canvas (`2365de6c`, `f10c84a9`), but Canvas had its own GPU context limits. Eventually both were removed (`7bd372e7`) and we landed on the pure DOM renderer for reliability.

## Dependency blocker

Here's the critical finding: `@xterm/addon-webgl@0.19.0` (the only stable release targeting xterm 6.x) declares a peer dependency of `@xterm/xterm: ^6.1.0-beta.193`. We're on `@xterm/xterm@6.0.0`. There's no stable addon-webgl that's compatible with our xterm version.

Three options:
1. **Bump xterm to `6.1.0-beta.193`** (latest beta, 193 beta releases deep so probably close to stable)
2. **Install addon-webgl 0.19.0 with `--legacy-peer-deps`** and hope for the best
3. **Wait for xterm 6.1.0 stable**

I'd lean toward option 1 or 3. Option 2 is asking for trouble given the rendering bugs we've already fought.

## Why this time is different

The previous system tried to manage 4-16 WebGL contexts simultaneously with visibility tracking, grace periods, and eviction logic. That complexity was the source of most bugs. This issue proposes WebGL for **only the focused terminal** (1 context at steady state), which is a fundamentally different problem.

`TerminalRendererPolicy` already provides FOCUSED/VISIBLE/BACKGROUND tiers with hysteresis, and `ManagedTerminal.isFocused` gives us the exact trigger signal. No LRU, no budget calculations, no hardware detection needed.

## Recommended implementation

Based on what went wrong last time:

1. **Resolve the dependency first.** Nothing else matters until we have a compatible addon-webgl.
2. **Single WebGL context only.** One addon instance, transferred on focus change. No `WebGLContextManager`, no budgets.
3. **Dispose-and-recreate pattern.** The `WebglAddon` can't be reattached after disposal, so create a new instance each time focus shifts.
4. **Debounce with existing hysteresis.** Reuse `TIER_DOWNGRADE_HYSTERESIS_MS` (500ms) to avoid churn during rapid tab switching.
5. **Context loss = silent DOM fallback.** `onContextLoss` disposes the addon and doesn't retry until the next focus change. The exponential backoff recovery from before (`7e05c769`) added significant complexity and was still fragile.
6. **No grace periods.** When focus leaves, dispose immediately. With 1 context there's no eviction logic to worry about.

The old `WebGLContextManager` code is in git history at `f10c84a9` for reference, but the new implementation should be maybe 50-80 lines vs the previous 227. The `WebglAddon` constructor in 6.x takes a single optional `preserveDrawingBuffer` boolean (we don't need it), and `onContextLoss` works the same way.
